app/code/Magento/ConfigurableProduct/Model/Product/Type/Configurable.php

```diff
@@ -1290,13 +1290,19 @@ public function isPossibleBuyFromList($product)
      */
     public function getUsedProducts($product, $requiredAttributeIds = null)
     {
-        if (!$product->hasData($this->_usedProducts)) {
+        $requiredAttributeIds = $requiredAttributeIds ?: [];
+        array_unshift($requiredAttributeIds, $product->getId());
+
+        $cacheKey = $this->getUsedProductsCacheKey($requiredAttributeIds);
+        if (!$product->hasData($this->_usedProducts) || !isset($product->getData($this->_usedProducts)[$cacheKey])) {
             $collection = $this->getConfiguredUsedProductCollection($product, false, $requiredAttributeIds);
             $usedProducts = array_values($collection->getItems());
-            $product->setData($this->_usedProducts, $usedProducts);
+            $usedProductsCache = $product->getData($this->_usedProducts);
+            $usedProductsCache[$cacheKey] = $usedProducts;
+            $product->setData($this->_usedProducts, $usedProductsCache);
         }
 
-        return $product->getData($this->_usedProducts);
+        return $product->getData($this->_usedProducts)[$cacheKey];
```

Comment:
I believe that this method could be reverted to https://github.com/magento/magento2/commit/2487db1b#diff-d9457baecd561c1374a935c202ad3a4ebf82a9c26b80c486dd97ac44da81d383L1245-L1257 with some changes:
* use `$this->customerSession` instead of deprecated `getCustomerSession()`
* add `$requiredAttributeIds` into `loadUsedProducts` call

And refactoring `loadUsedProducts` method